### ISSUE TITLE
Manually vendored and ignored packages are removed during "glide install"  (0.12.0 only)

### ISSUE THREAD

**Issue by mandarjog** [Masterminds/glide#579]
https://github.com/Masterminds/glide/commit/9f8eca0c2806a5a5fed5c47a897932805040b815#commitcomment-18762612

This bug is introduced in 0.12.0 
v0.11.1 works fine.

I had vendored some generated code in vendor/ directory and listed the packages as "ignore" in yaml file.
0.11.1 correctly leaves those alone as they are actually checked in to VCS. 
In 0.12.0 those files are deleted

**Comment by nxvl** [comment]
Was going to report the same issue. I use the vendored packages for:
- packages that need manual building (bindata packages)
- private repos (cause heroku doesn't handle them correctly)
since 0.12 my build has got broken.

**Comment by mattfarina** [comment]
First, let me cc @sdboyer as he needs to be in the loop.

Glide's intent was to manage your vendor folder. We'd not considered having manually vendored packages alongside Glide managed ones. Manually vendoring means nothing has the true recorded state of the things (which is what a lockfile, like `glide.lock`).

Feedback, discussion, and suggestions are very welcome. I'm pragmatic.

**Comment by mandarjog** [comment]
Here is the scenario.

I have several deeply nested protobuf definitions. I can generate golang code from that. The generated code uses non absolute imports from within the protobuf tree. 

I can periodically generate and checkin the generated code as a vendored package.

I can keep metadata associate with it out of band.
Is there a better way to do it?

Regards, 
Sent from my Ericsson Blaze

Mandar U Jog

**Comment by sdboyer** [comment]
Yeah, I've spent a while thinking about the best handling for this case, but until people actually came up with examples where it was the _only_ option, I deferred action on it.

I think that the key question, @mandarjog and @nxvl, is whether these packages that you're currently storing under `vendor/` _have_ to be stored there, or if they can be placed under the local package tree?

If so, is it considerably more difficult to place them under the local tree than under `vendor/`?

> The generated code uses non absolute imports from within the protobuf tree.

Just to be clear - this means it's using relative imports, like `../protob/some/path`?

**Comment by mattfarina** [comment]
@mandarjog I'm sorry I didn't figure some folks would be putting generated code in the `vendor/` directory.

The reason for that is the intent of the `vendor/` directory is for [_external_ dependencies](https://docs.google.com/document/d/1Bz5-UB7g2uPBdOx-rw5t9MxJwkfpx90cqG9AFL0JAYo/edit). Generated code that's for your application isn't an external dependency. Protocol buffers that are part of another application or library should have their generated source checked in and that external package can be imported and is external so it fits in `vendor/`. For example, [the Go blog notes of generated code](https://blog.golang.org/generate):

> it must be checked into the source code repository to be available to clients

I had not anticipated generated code being in `vendor/` and this is the first I've heard of someone doing it.

**Comment by mattfarina** [comment]
@nxvl I may be able to address your comments which are a little different from the original.

* Glide can handle private packages using the `repo` property. See the [glide.yaml file docs](https://glide.readthedocs.io/en/latest/glide.yaml/). You can specify a repo, such as `git@git.example.com/foo/bar.git`. If you're using a build environment that does not know how to access your private packages check them into the app.
* The packages that need manual building.... `vendor/` was designed for external dependency storage rather than our internally generated elements. In writing responses here I double checked the intent of `vendor/` from the original authors. This may not be the best place for your built elements.

**Comment by nxvl** [comment]
@mattfarina 
- Yes, i indeed do and that works great when i build my package in circleci, but, when deploying to heroku, since they do not have deploy keys (unless i go to overhead of adding an extra buildpack and configure it) my private repos can't be fetched.
- What i'm vendoring is https://github.com/goadapp/goad they require `make binutils` to be run to create the `infrastructure/bindata.go` file. If that file is not created the build will fail since Asset is not declared. So i manually vendor it to remove that file from gitignore and check it into my repo.

**Comment by sdboyer** [comment]
> they require make binutils to be run to create the infrastructure/bindata.go file. If that file is not created the build will fail since Asset is not declared. So i manually vendor it to remove that file from gitignore and check it into my repo.

That's...nasty. They require code depending on their code to generate something _within_ their code tree in order for it to work correctly? Yikes. Yeah, gotta think about that one.

**Comment by nxvl** [comment]
Yup, and reported an issue letting them know that too: https://github.com/goadapp/goad/issues/32

**Comment by sdboyer** [comment]
@nxvl hmm...I read a little more about goad - what's wrong with just running their `make binutils` as part of _your_ CI process, if it's something you need?

**Comment by nxvl** [comment]
@sdboyer if i `cd vendor/.../goad && make binutils` go complains because it's not on the gopath and some dependencies are missing. But if i do `go get github..../goad && cd .../goad && make binutils` works (which is what i do and then copy the missing file.)

**Comment by bakks** [comment]
Chiming in here because the behavior in `0.12.0` was unexpected and just bit me. I'd vote in the direction of glide leaving existing `vendor/*` directories intact. I have a similar protobuf use-case as mentioned above, where I'm generating Go code from a service definition, but want to keep it specific and isolated to one of my packages. Yes, I could generate the code and and put it in the top-level package tree, but what if I want separate versions in each of my packages? This is the whole point of the vendor directory, as I understand it.

Basically - glide seems less useful if it can't live alongside other vendored code.

Also - at the very least I think glide should warn the user before deleting manually vendored code, this behavior was super unexpected.

Hope this is constructive, thanks for talking through this.

**Comment by yaa110** [comment]
I started a project with Glide, then created some local packages inside vendor folder (which makes it possible to change the path of project without worrying about GOPATH), I wrote some codes, before staging them, I ran glide install ... oops! ... my code has gone.  I marked those local packages as ignore in the manifest file, but it seems glide does not respect to its features. So, I decided to write my own package manager which respects to my manual changes: https://github.com/yaa110/rubigo
